tools/autograd/gen_annotated_fn_args.py

```diff
@@ -13,65 +13,69 @@
-
-    for func in recurse_dict(get_py_nn_functions(declarations)):
-        annotated_args.append(process_func("torch._C._nn", func))
-
-    for func in recurse_dict(get_py_variable_methods(declarations)):
-        annotated_args.append(process_func("torch.Tensor", func))
-
-    annotated_args = textwrap.indent("\n".join(annotated_args), "    ")
-    env = {"annotated_args": annotated_args}
-    PY_ANNOTATED_ARGS = CodeTemplate.from_file(template_path + '/templates/annotated_fn_args.py')
-    write(out, 'annotated_fn_args.py', PY_ANNOTATED_ARGS, env)
-
-
-def process_func(namespace, func):
-    args = func["arguments"]
-    out_args = []
-    for arg in args:
-        if 'default' in arg or arg.get('kwarg_only', False) or arg.get('output', False):
+from typing import Dict, List, Any
+
+from tools.codegen.gen import with_native_function, parse_native_yaml, FileManager
+from tools.codegen.model import *
+import tools.codegen.api.python as python
+from .gen_python_functions import should_generate_py_binding, is_py_torch_function, is_py_nn_function, is_py_variable_method
+
+def gen_annotated(native_yaml_path: str, out: str, autograd_dir: str) -> None:
+    native_functions = parse_native_yaml(native_yaml_path)
+    mappings = (
+        (is_py_torch_function, 'torch._C._VariableFunctions'),
+        (is_py_nn_function, 'torch._C._nn'),
+        (is_py_variable_method, 'torch.Tensor'),
+    )
+    annotated_args: List[str] = []
+    for pred, namespace in mappings:
+        groups: Dict[str, List[NativeFunction]] = defaultdict(list)
+        for f in native_functions:
+            if not should_generate_py_binding(f) or not pred(f):
+                continue
+            groups[str(f.func.name.name)].append(f)
+        for group in groups.values():
+            for f in group:
+                annotated_args.append(f'{namespace}.{process_func(f)}')
+
+    template_path = os.path.join(autograd_dir, 'templates')
+    fm = FileManager(install_dir=out, template_dir=template_path, dry_run=False)
+    fm.write_with_template('annotated_fn_args.py', 'annotated_fn_args.py', lambda: {
+        'annotated_args': textwrap.indent('\n'.join(annotated_args), '    '),
+    })
+
+@with_native_function
+def process_func(f: NativeFunction) -> str:
+    out_args: List[Dict[str, Any]] = []
+    for arg in f.func.arguments:
+        if arg.default is not None:
             continue
-        out_args.append({k: arg[k] for k in ('name', 'simple_type', 'size') if k in arg})
-
-    return f"{namespace}.{op_name(func)}: {out_args!r},"
-
+        out_arg: Dict[str, Any] = {}
+        out_arg['name'] = arg.name
+        out_arg['simple_type'] = python.argument_type_str(arg.type, simple_type=True)
+        size = python.argument_type_size(arg.type)
+        if size:
+            out_arg['size'] = size
+        out_args.append(out_arg)
 
-def recurse_dict(d):
-    for e in d.values():
-        for i in e:
-            yield i
+    return f'{f.func.name.name}: {repr(out_args)},'
```

Comment:
Oh wait, the func.names here have namespaces attached to them?

Reply:
> Oh wait, the func.names here have namespaces attached to them?

hmm, I'm not sure I understand? you mean 'aten::' namespace or 'torch._C._VariableFunctions' namespace?
We don't have the former in the output file. The latter is added explicitly above.